### PR TITLE
Add Wheel Input into Webdriver

### PR DESCRIPTION
**Pull request by LanWei22**
Add Wheel Input into Webdriver's implementation testdriver Action API and add Webdriver wheel tests

### CODE REVIEW

**Comment by SummerLW** [comment]
David and James, I added the implementation to testdriver and a Webdriver wheel test, please take a look, thank you.

File: tools/webdriver/webdriver/client.py [hunk 1]
```diff
@@ -203,6 +203,32 @@ def send_keys(self, keys):
...
+            "y": y,
+            "deltaX": delta_x,
+            "deltaY": delta_y
+        }
+        if duration is not None:
+            action["duration"] = duration
+        if origin is not None:
+            action["origin"] = origin
+        self._actions.append(action)
+        return self
```
Comment (LukeZielinski) [inline tools/webdriver/webdriver/client.py:230]: This line has trailing whitespace which is causing lint failures

**Review by LukeZielinski — COMMENTED** [review]
The wpt-chrome-dev-stability failures are a bit odd. The stability check claims this test has a "duplicate subtest name", which doesn't make sense. The logs are also showing setup/teardown errors for this test, as well as chromedriver response errors.

Does this test work locally? Does it work with --verify as well?

**Comment by LanWei22** [comment]
LukeZielinski@, I fixed the flaky test,  now there is still some error, do you think it is from the infrastructure?

**Comment by LukeZielinski** [comment]
Yeah the failure in tools/wpt tests on windows seem to be infra issues.

**Review by LukeZielinski — APPROVED** [review]

File: resources/testdriver-actions.js [hunk 2]
```diff
@@ -154,6 +155,32 @@
...
+     * Add a new wheel input source with the given name
+     *
+     * @param {String} type - Name of the wheel source
+     * @param {Bool} set - Set source as the default wheel source
+     * @returns {Actions}
+     */
+    addWheel: function(name, set=true) {
+      this.createSource("wheel", name);
+      if (set) {
+        this.setPointer(name);
```
Comment (jgraham) [inline resources/testdriver-actions.js:168]: `setPointer` here seems wrong?

File: resources/testdriver-actions.js [hunk 3]
```diff
@@ -417,5 +466,46 @@
     },
   };
 
+  function WheelSource() {
```
Comment (jgraham) [inline resources/testdriver-actions.js:469]: I'd like to see a test under `infrastructure` that uses the wheel source, as a sanity check.

File: webdriver/tests/perform_actions/pointer.py [hunk 4]
```diff
@@ -148,7 +148,7 @@ def test_drag_and_drop_with_draggable_element(session_new_window,
                        test_actions_page,
                        mouse_chain,
                        drag_duration):
-    new_session = session_with_new_window
+    new_session = session_new_window
```
Comment (jgraham) [inline webdriver/tests/perform_actions/pointer.py:151]: Why this change?

File: webdriver/tests/perform_actions/wheel.py [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+def test_no_browsing_context(session, closed_window, wheel_chain):
+    with pytest.raises(NoSuchWindowException):
+        wheel_chain.scroll(0, 0, 0, 10).perform()
+
+
+def test_wheel_scroll(session, test_actions_page, wheel_chain):
+    session.execute_script("document.scrollingElement.scrollTop = 0")
+    div_point = { "x": 82, "y": 187}
+
+    wheel_chain \
```
Comment (jgraham) [inline webdriver/tests/perform_actions/wheel.py:23]: In general I don't love the continutation lines here.

File: webdriver/tests/perform_actions/wheel.py [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+    value = wheel_chain.scroll(0, 0, 0, 10).perform()
+    assert value is None
+
+
+def test_no_browsing_context(session, closed_window, wheel_chain):
+    with pytest.raises(NoSuchWindowException):
+        wheel_chain.scroll(0, 0, 0, 10).perform()
+
+
+def test_wheel_scroll(session, test_actions_page, wheel_chain):
```
Comment (jgraham) [inline webdriver/tests/perform_actions/wheel.py:19]: Would be good to also have a test for scrolling something other than the main document e.g. an iframe or something with overflow: scroll.

**Review by jgraham — DISMISSED** [review]

File: webdriver/tests/perform_actions/pointer.py (hunk 4 above)
Comment (LanWei22) [reply webdriver/tests/perform_actions/pointer.py:151]: This is from my other fix https://github.com/web-platform-tests/wpt/pull/25365, I merged it. I do not know how to rebase this PR on top of it.

**Comment by LanWei22** [comment]
jgraham@ I have fixed all the errors, please take a look, thank you.

File: webdriver/tests/perform_actions/wheel.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
+import pytest
+
+from webdriver.error import NoSuchWindowException
+
+from tests.perform_actions.support.refine import filter_dict, get_events
+from tests.support.asserts import assert_move_to_coordinates
+
+
+@pytest.fixture(autouse=True)
+def release_actions(session, request):
```
Comment (LanWei22) [inline webdriver/tests/perform_actions/wheel.py:10]: The reason I overwrote it here is this fails right now, but I have fixed it here https://chromium-review.googlesource.com/c/chromium/src/+/2387546, which has not been released to the chromedriver version that is used to run this PR, but it should be released soon.

**Comment by whimboo** [comment]
@LanWei22 can you please check the other PRs you have created - most likely accidentally? I requested feedback from you on those. I assume we can get those closed. Thanks.

**Comment by LanWei22** [comment]
whimboo@ yes, I created the other two to test some files, because this PR has too many files, it had some errors before, so I have to break it into small pieces to test it. Now it passes all checks, I will delete the other files.

**Comment by stephenmcgruer** [comment]
> whimboo@ yes, I created the other two to test some files, because this PR has too many files, it had some errors before, so I have to break it into small pieces to test it. Now it passes all checks, I will delete the other files.

As a note, if you create PRs as drafts (step 6 of https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request), the tests will run but nobody will be automatically assigned and cc'd . Very useful for testing changes :)

**Comment by LanWei22** [comment]
> > whimboo@ yes, I created the other two to test some files, because this PR has too many files, it had some errors before, so I have to break it into small pieces to test it. Now it passes all checks, I will delete the other files.
> 
> As a note, if you create PRs as drafts (step 6 of https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request), the tests will run but nobody will be automatically assigned and cc'd . Very useful for testing changes :)
> 
> Thank you!



> > whimboo@ yes, I created the other two to test some files, because this PR has too many files, it had some errors before, so I have to break it into small pieces to test it. Now it passes all checks, I will delete the other files.
> 
> As a note, if you create PRs as drafts (step 6 of https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/creating-a-pull-request), the tests will run but nobody will be automatically assigned and cc'd . Very useful for testing changes :)

Thank you!

**Comment by stephenmcgruer** [comment]
Please wait 24 hours for any comments from @jgraham - if there are no further concerns this can be merged.

File: webdriver/tests/perform_actions/pointer.py (hunk 4 above)
Comment (whimboo) [reply webdriver/tests/perform_actions/pointer.py:151]: You want to pull the latest changes from the upstream repository and then run a `git rebase master` while having your branch for wheel input checked out.

File: webdriver/tests/perform_actions/wheel.py [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+from tests.perform_actions.support.refine import filter_dict, get_events
+from tests.support.asserts import assert_move_to_coordinates
+
+
+def test_null_response_value(session, wheel_chain):
+    value = wheel_chain.scroll(0, 0, 0, 10).perform()
+    assert value is None
+
+
+def test_no_browsing_context(session, closed_window, wheel_chain):
```
Comment (whimboo) [inline webdriver/tests/perform_actions/wheel.py:14]: Please add another test for the top-level browsing context similar to what I have done here: https://hg.mozilla.org/integration/autoland/rev/12fadd0009e4. These changes will get a new PR today and getting merged soon by @jgraham. I simply don't want that we miss to add the appropriate check.

File: webdriver/tests/perform_actions/wheel.py [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        wheel_chain.scroll(0, 0, 0, 10).perform()
+
+
+def test_wheel_scroll(session, test_actions_scroll_page, wheel_chain):
+    session.execute_script("document.scrollingElement.scrollTop = 0")
+
+    outer = session.find.css("#outer", all=False)
+    wheel_chain.scroll(0, 0, 5, 10, origin=outer).perform()
+    events = get_events(session)
+    assert len(events) > 0
```
Comment (whimboo) [inline webdriver/tests/perform_actions/wheel.py:25]: Do we expect to see a different number of events here? It would be good to have an equal check instead. This also applies to the other tests in this file.

File: webdriver/tests/perform_actions/wheel.py [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+    event_types = [e["type"] for e in events]
```
Comment (whimboo) [inline webdriver/tests/perform_actions/wheel.py:26]: Why don't we just filter for `type` here? This would prevent us from having the additional for loop. This also applies to the other tests in this file.

File: webdriver/tests/perform_actions/wheel.py [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+            assert e["deltaY"] >= 10
+            assert e["deltaZ"] == 0
+            assert e["target"] == "scrollContent"
+
+
+def test_wheel_scroll_iframe(session, test_actions_scroll_page, wheel_chain):
+    session.execute_script("document.scrollingElement.scrollTop = 0")
+
+    subframe = session.find.css("#subframe", all=False)
+    wheel_chain.scroll(0, 0, 5, 10, origin=subframe).perform()
```
Comment (whimboo) [inline webdriver/tests/perform_actions/wheel.py:57]: So that means we operate on a sub-frame while not having switched to it before? From what I can see that shouldn't work, or is wheel scrolling different here? @jgraham?

**Review by whimboo — DISMISSED** [review]

File: webdriver/tests/perform_actions/wheel.py (hunk 9 above)
Comment (LanWei22) [reply webdriver/tests/perform_actions/wheel.py:25]: Yes, we do not know for sure how many events we will get because this test may get a mouse move which depends on the current mouse position.

**Comment by cbiesinger** [comment]
I'm not a good reviewer for this. Also, a lot of those changes seem unrelated?

**Comment by stephenmcgruer** [comment]
> I'm not a good reviewer for this. Also, a lot of those changes seem unrelated?

I think there has been a git-rebase accident. Apologies to those that were auto-cc'd due to the unrelated changes.

@LanWei22 - feel free to ping me sometime today and I can help you figure out what has happened here and get this PR back to an expected state :)

File: webdriver/tests/perform_actions/wheel.py (hunk 11 above)
Comment (LanWei22) [reply webdriver/tests/perform_actions/wheel.py:56]: I have submitted a CL, which switched the frame https://github.com/web-platform-tests/wpt/blob/e949f1f611a601e82dca420c26d092fdd3475155/tools/webdriver/webdriver/client.py#L326, if the element is an iframe.

File: webdriver/tests/perform_actions/wheel.py (hunk 8 above)
Comment (LanWei22) [reply webdriver/tests/perform_actions/wheel.py:19]: Do you mean "test_no_top_browsing_context"? I added this test.

**Review by LanWei22 — COMMENTED** [review]
jgraham@ and whimboo@, please take a look, thank you.

**Comment by stephenmcgruer** [comment]
Looks like the reviewers were not pinged with correct GitHub syntax (@ in front of name), so pinging @jgraham and @whimboo in case they have comments.

@LukeZielinski - you originally reviewed this, but it has changed a bit since then. Can you take another look please?

@LanWei22 - once @LukeZielinski has taken another look and LGTM'd, please give the other two 24 hours for comment then merge. Thanks!

File: webdriver/tests/perform_actions/wheel.py (hunk 10 above)
Comment (LanWei22) [reply webdriver/tests/perform_actions/wheel.py]: Done.

**Review by LukeZielinski — APPROVED** [review]
Still lgtm from me

File: webdriver/tests/perform_actions/wheel.py (hunk 9 above)
Comment (whimboo) [reply webdriver/tests/perform_actions/wheel.py]: Please do so. That way we can assert for a clear number of wheel events.

File: webdriver/tests/perform_actions/wheel.py (hunk 11 above)
Comment (whimboo) [reply webdriver/tests/perform_actions/wheel.py:56]: So this was added by https://github.com/web-platform-tests/wpt/pull/17260 for finding an element via a CSS selector. Is that really something we are allowed to do? In my opinion that shouldn't have been merged. Nothing in `Find Element` should switch the frame. @jgraham please give your feedback here, which might be fixed in a follow-up PR.

Comment (whimboo) [reply webdriver/tests/perform_actions/wheel.py:56]: Oh wait. It's an additional option for the Python client. So it should be fine, but still hides frame switches in the tests, which why I was confused. Having to explicitly switch to a frame in the test should be better.
